File: lista.1/batalha.naval/Tabuleiro.java
import java.util.Scanner;

class Tabuleiro {
  private char[][] tabuleiro;
  private int tamanho;
  private int numNavios;

  public Tabuleiro(int tamanho, int numNavios) {
    this.tamanho = tamanho;
    this.numNavios = numNavios;
    this.tabuleiro = new char[tamanho][tamanho];

    for (int i = 0; i < tamanho; i++) {
      for (int j = 0; j < tamanho; j++) {
        tabuleiro[i][j] = '-';
      }
    }
  }

  public void posicionarNavios() {
    Scanner scanner = new Scanner(System.in);

    for (int i = 0; i < numNavios; i++) {
      System.out.println("Enter ship " + (i + 1) + " location:");
      int linha, coluna;
      do {
        linha = scanner.nextInt();
        coluna = scanner.nextInt();
      } while (!posicionarNavio(linha, coluna));
    }

    this.exibir(true);
  }

  private boolean posicionarNavio(int linha, int coluna) {
    if (linha > 4 || coluna > 4 || linha < 0 || coluna < 0) {
      System.out.println("Invalid coordinates. Choose different coordinates.");
      return false;
    } else if (tabuleiro[linha][coluna] == '@') {
      System.out.println("Invalid coordinates. Choose different coordinates.");
      return false;
    }

    tabuleiro[linha][coluna] = '@';
    return true;
  }

  public boolean atirar(int linha, int coluna, String nome) {
    if (linha > 4 || coluna > 4 || linha < 0 || coluna < 0) {
      System.out.println("Invalid coordinates. Choose different coordinates.");
      return false;
    } else if (tabuleiro[linha][coluna] == '@') {
      tabuleiro[linha][coluna] = 'X';
      System.out.println(nome + " HIT SHIP!\n");
      return true;
    } else {
      tabuleiro[linha][coluna] = 'O';
      System.out.println(nome + " MISSED!");
      return true;
    } 
  }

  public boolean perdeu() {
    for (int i = 0; i < tamanho; i++) {
      for (int j = 0; j < tamanho; j++) {
        if (tabuleiro[i][j] == '@') {
          return false;
        }
      }
    }
    return true;
  }

  public void exibir(boolean mostrarNavios) {
    System.out.println("  0 1 2 3 4");
    for (int i = 0; i < tamanho; i++) {
      System.out.print(i + " ");
      for (int j = 0; j < tamanho; j++) {
        if (!mostrarNavios && tabuleiro[i][j] == '@') {
          System.out.print("- ");
        } else {
          System.out.print(tabuleiro[i][j] + " ");
        }
      }
      System.out.println();
    }
    System.out.println();
  }
}

File: lista.2/batalha.naval/Tabuleiro.py
class Tabuleiro:
    
    def __init__(self, tamanho, numNavios):
        self.tamanho = tamanho
        self.numNavios = numNavios
        self.tabuleiro = [['-','-','-','-','-'],
                          ['-','-','-','-','-'],
                          ['-','-','-','-','-'],
                          ['-','-','-','-','-'],
                          ['-','-','-','-','-'],]

    def posicionarNavios(self):
        for i in range(0, self.tamanho):
            print(f"Enter ship {(i + 1)} location:")

            while True:
                linha = int(input())
                coluna = int(input())

                if self.__posicionarNavio(linha, coluna):
                    break
        
        self.exibir(True)

    def __posicionarNavio(self, linha, coluna):
        if linha > 4 or coluna > 4 or linha < 0 or coluna < 0:
            print("Invalid coordinates. Choose different coordinates.")
            return False
        elif self.tabuleiro[linha][coluna] == '@':
            print("Invalid coordinates. Choose different coordinates.")
            return False
        
        self.tabuleiro[linha][coluna] = '@'
        return True;

    def atirar(self, linha, coluna, nome):
        if linha > 4 or coluna > 4 or linha < 0 or coluna < 0:
            print("Invalid coordinates. Choose different coordinates.")
            return False
        elif self.tabuleiro[linha][coluna] == '@':
            self.tabuleiro[linha][coluna] = 'X'
            print(f"{nome} HIT SHIP!")
            return True
        else:
            self.tabuleiro[linha][coluna] = 'O'
            print(f"{nome} MISSED!")
            return True
        
    def perdeu(self):
        for i in range(0, self.tamanho):
            for j in range(0, self.tamanho):
                if self.tabuleiro[i][j] == '@':
                    return False
        return True

    def exibir(self, mostrarNavios):
        print("  0 1 2 3 4")
        for i in range(0, self.tamanho):
            print(f"{i} ", end="")
            for j in range(0, self.tamanho):
                if not mostrarNavios and self.tabuleiro[i][j] == '@':
                    print("- ", end="")
                else: 
                    print(f"{self.tabuleiro[i][j]} ", end="")
            print()
        print()